File: TwoPointers/threeSum.ts
function threeSum(nums: number[]): number[][] {
    let sortNums = nums.sort((a, b) => a - b);
    let trip: number[][] = [];
    for (let i = 0; i < sortNums.length - 1; i++) {
        let y = i + 1;
        let z = sortNums.length - 1;
        if (i > 0 && sortNums[i] === sortNums[i - 1]) {
            continue;
        }
        while (y < z) {
            let sum = sortNums[i] + sortNums[y] + sortNums[z];
            if (sum === 0) {
                trip.push([sortNums[i], sortNums[y], sortNums[z]]);
                y++;
                while (nums[y] !== nums[y - 1] && y < z) {
                    y++;
                }
            } else {
                sum < 0 ? y++ : z--;
            }
        }
    }
    return trip;
}

File: TwoPointers/three_sum.cpp
#include <algorithm>
#include <cstdlib>
#include <map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int> &nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        int a = 0;

        while (a < nums.size()) {
            int b = a + 1;
            int c = nums.size() - 1;
            while (b < c) {
                if (nums[a] + nums[b] + nums[c] > 0) {
                    c--;
                } else if (nums[a] + nums[b] + nums[c] < 0) {
                    b++;
                } else {
                    vector<int> trip = {nums[a], nums[b], nums[c]};
                    if (a != b && b != c && a != c) {
                        ans.push_back(trip);
                    }
                    c--;
                    b++;
                }
            }
        }
        return ans;
    }
};
